### PR TITLE
# 231: decorators can't read token from cookie

### LINKED ISSUE #231: Decorators don't read token from cookies
auth decorators only try reading a token from the header, not cookies

### PR DESCRIPTION
**Pull request by jvllmr**
This is a rewrite of PR #255 . Without the enum feature, with less commits and extended testing.

Note that I have reformatted the edited files with black.

fixes #231

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -230,3 +230,30 @@ def clean_flask_app_config(app):
         stock_config = app.config.copy()
         yield
         app.config = stock_config.copy()
+
+
+@pytest.fixture(autouse=True)
+def use_cookie(client, default_guard):
+    class withCookie:
```
Comment (dusktreader) [inline tests/conftest.py:238]: 🔥

File: tests/test_decorators.py [hunk 2]
```diff
@@ -52,90 +50,84 @@ def test_auth_accepted(self, client, default_guard):
...
-            assert response.json['user'] == self.the_dude.username
+            assert "success" in response.json["message"]
+            assert response.json["user"] == self.the_dude.username
 
-    def test_auth_required(self, client, default_guard):
+    def test_auth_required(self, client, default_guard, use_cookie):
         """
         This test verifies that the @auth_required decorator can be used
         to ensure that any access to a protected endpoint must have a properly
         structured auth header including a valid jwt token.  Otherwise,
```
Comment (dusktreader) [inline tests/test_decorators.py]: Please update docstring to mention that it also covers the case of a token in the cookie.

File: flask_praetorian/base.py [hunk 3]
```diff
@@ -296,18 +296,22 @@ class methods and attributes necessary to function correctly.
...
         PraetorianError.require_condition(
-            getattr(user_class, 'lookup', None) is not None,
-            textwrap.dedent("""
+            getattr(user_class, "lookup", None) is not None,
+            textwrap.dedent(
+                """
                 The user_class must have a lookup class method:
                 user_class.lookup(<str>) -> <user instance>
-            """),
+            """
```
Comment (dusktreader) [inline flask_praetorian/base.py]: please indent this

File: flask_praetorian/base.py [hunk 4]
```diff
@@ -296,18 +296,22 @@ class methods and attributes necessary to function correctly.
...
         PraetorianError.require_condition(
-            getattr(user_class, 'identify', None) is not None,
-            textwrap.dedent("""
+            getattr(user_class, "identify", None) is not None,
+            textwrap.dedent(
+                """
                 The user_class must have an identify class method:
                 user_class.identify(<identity>) -> <user instance>
-            """),
+            """
```
Comment (dusktreader) [inline flask_praetorian/base.py]: please indent this.

File: flask_praetorian/base.py [hunk 5]
```diff
@@ -321,24 +325,30 @@ class methods and attributes necessary to function correctly.
         if dummy_user:
             PraetorianError.require_condition(
                 hasattr(dummy_user, "identity"),
-                textwrap.dedent("""
+                textwrap.dedent(
+                    """
                     Instances of user_class must have an identity attribute:
                     user_instance.identity -> <unique id for instance>
-                """),
+                """
```
Comment (dusktreader) [inline flask_praetorian/base.py]: please indent this.

File: flask_praetorian/base.py [hunk 6]
```diff
@@ -321,24 +325,30 @@ class methods and attributes necessary to function correctly.
...
             )
             PraetorianError.require_condition(
                 self.roles_disabled or hasattr(dummy_user, "rolenames"),
-                textwrap.dedent("""
+                textwrap.dedent(
+                    """
                     Instances of user_class must have a rolenames attribute:
                     user_instance.rolenames -> [<role1>, <role2>, ...]
-                """),
+                """
```
Comment (dusktreader) [inline flask_praetorian/base.py]: please indent this (and all following cases)

File: flask_praetorian/base.py [hunk 7]
```diff
@@ -721,25 +740,46 @@ def read_token_from_cookie(self):
...
+                    self,
+                    "read_token_from_{place}".format(
+                        place=place.lower(),
+                    ),
+                )()
+            except MissingToken:
+                pass
+            except AttributeError:
+                flask.current_app.logger.warning(
+                    "Flask_Praetorian hasn't implemented reading JWT tokens from location {place}. Please reconfigure JWT_PLACES.".format(  # noqa: E501
```
Comment (dusktreader) [inline flask_praetorian/base.py]: This is a good place to use dedent:
```suggestion
                    snick.dedent(
                        f"""
                        Flask_Praetorian hasn't implemented reading JWT tokens from location {place.lower}.
                        Please reconfigure JWT_PLACES.
                        """
                    )
```

File: flask_praetorian/base.py [hunk 8]
```diff
@@ -721,25 +740,46 @@ def read_token_from_cookie(self):
...
+                        place=place.lower(),
+                    )
+                )
+                flask.current_app.logger.warning(
```
Comment (dusktreader) [inline flask_praetorian/base.py]: this logger warning should be folded into the previous one. This way, if it is captured by log aggregators in a concurrent webserver setup, the messages appear together.

File: flask_praetorian/base.py [hunk 9]
```diff
@@ -721,25 +740,46 @@ def read_token_from_cookie(self):
...
+                    "Values accepted in JWT_PLACES are: {def_places}".format(
+                        def_places=DEFAULT_JWT_PLACES,
+                    )
+                )
+
+        raise MissingToken(
+            "Could not find token in any of the given locations: {locations}".format(  # noqa: E501
```
Comment (dusktreader) [inline flask_praetorian/base.py]: use dedent if the line is too long.

**Review by dusktreader — CHANGES_REQUESTED** [review]
This is great work! Thank you. Left a few comments for you, but overall it looks good.

File: tests/conftest.py (hunk 1 above)
Comment (jvllmr) [reply tests/conftest.py:238]: Either you're telling me that there is somethign wrong or that this part is just lit. I'll gladly take it as the last one 😅

File: flask_praetorian/base.py (hunk 3 above)
Comment (jvllmr) [reply flask_praetorian/base.py]: indented now

File: tests/test_decorators.py (hunk 2 above)
Comment (jvllmr) [reply tests/test_decorators.py]: updated.

File: flask_praetorian/base.py (hunk 4 above)
Comment (jvllmr) [reply flask_praetorian/base.py]: indented now

File: flask_praetorian/base.py (hunk 5 above)
Comment (jvllmr) [reply flask_praetorian/base.py]: indented now

File: flask_praetorian/base.py (hunk 6 above)
Comment (jvllmr) [reply flask_praetorian/base.py]: I hope I didn't leave one out

File: flask_praetorian/base.py (hunk 7 above)
Comment (jvllmr) [reply flask_praetorian/base.py]: used the alreay imported texwrap module for this

File: flask_praetorian/base.py (hunk 8 above)
Comment (jvllmr) [reply flask_praetorian/base.py]: merged

File: flask_praetorian/base.py (hunk 9 above)
Comment (jvllmr) [reply flask_praetorian/base.py]: used dedent for this, but had to use .replace("\n", "") at the end to avoid \n inconsistencies

Comment (jvllmr) [reply flask_praetorian/base.py]: the inconsistencies make the tests fail. uh oh

Comment (jvllmr) [reply flask_praetorian/base.py]: a tab sneaked in somehow after I ran the tests on my machine... the tests did run through now

File: tests/conftest.py (hunk 1 above)
Comment (dusktreader) [reply tests/conftest.py:238]: Yeah, it's the latter. I like it

**Review by dusktreader — APPROVED** [review]

**Comment by dusktreader** [comment]
Everything looks good, but seems like the branch needs to be rebased on master.

**Comment by jvllmr** [comment]
Branch got rebased

**Pull request merged**